qiita_client/plugin.py

```diff
@@ -0,0 +1,330 @@
+            # information, the QiitaClient already raises an error
+            job_info = qclient.get_job_info(job_id)
+            # Starting the heartbeat
+            qclient.start_heartbeat(job_id)
+            # Execute the given task
+            task_name = job_info['command']
+            task = self.task_dict[task_name]
+
+            if not exists(output_dir):
+                makedirs(output_dir)
+            try:
+                success, artifacts_info, error_msg = task(
+                    qclient, job_id, job_info['parameters'], output_dir)
+            except Exception:
+                exc_str = repr(traceback.format_exception(*sys.exc_info()))
+                error_msg = ("Error executing %s:\n%s" % (task_name, exc_str))
+                success = False
+                artifacts_info = None
+            # The job completed
+            qclient.complete_job(job_id, success, error_msg=error_msg,
+                                 artifacts_info=artifacts_info)
+
+
+class QiitaTypePlugin(BaseQiitaPlugin):
+    """Represents a Qiita Type Plugin
+
+    Parameters
+    ----------
+    name : str
+        The plugin name
+    version : str
+        The plugin version
+    description : str
+        The plugin description
+    validate_func : callable
+        The function used to validate artifacts
+    html_generator_func : callable
+        The function used to generate the HTML generator
+    artifact_types : list of QiitaArtifactType
+        The artifact types defined in this plugin
+
+    Notes
+    -----
+    Both `validate_func` and `html_generator_func` should be a callable
+    that conforms to the signature:
+    function(qclient, job_id, job_parameters, output_dir)
+    where qclient is an instance of QiitaClient, job_id is a string with
+    the job identifier, job_parameters is a dictionary with the parameters
+    of the command and output_dir is a string with the output directory
+    """
+    _plugin_type = "artifact definition"
+
+    def __init__(self, name, version, description, validate_func,
+                 html_generator_func, artifact_types, publications=None):
+        super(QiitaTypePlugin, self).__init__(name, version, description,
+                                              publications=publications)
+
+        self.artifact_types = artifact_types
+
+        val_cmd = QiitaCommand(
+            'Validate', 'Validates a new artifact', validate_func,
+            {'template': ('prep_template', None),
+             'files': ('string', None),
+             'artifact_type': ('string', None)}, {}, None)
+
+        self._register_command(val_cmd)
+
+        html_cmd = QiitaCommand(
+            'Generate HTML summary', 'Generates the HTML summary',
+            html_generator_func,
+            {'input_data': ('artifact',
+                            [a.name for a in self.artifact_types])}, {}, None)
+
+        self._register_command(html_cmd)
+
+    def _install(self, qclient):
```

Comment:
it is getting executed here: https://github.com/qiita-spots/qiita_client/pull/8/files#diff-259ba12727397257087e45ba8a60f40dR163